httptransport/client/indexer.go

```diff
@@ -59,36 +65,40 @@ func (s *HTTP) AffectedManifests(ctx context.Context, v []claircore.Vulnerabilit
 // could not start. If an error occurs during the index operation the error will
 // be preset on the IndexReport.Err field of the returned IndexReport.
 func (s *HTTP) Index(ctx context.Context, manifest *claircore.Manifest) (*claircore.IndexReport, error) {
-	buf := bytes.NewBuffer([]byte{})
-	err := json.NewEncoder(buf).Encode(manifest)
-	if err != nil {
-		return nil, &clairerror.ErrBadManifest{err}
-	}
-
 	u, err := s.addr.Parse(httptransport.IndexAPIPath)
 	if err != nil {
 		return nil, fmt.Errorf("failed to create request: %v", err)
 	}
-	req, err := http.NewRequestWithContext(ctx, "POST", u.String(), buf)
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodPost, u.String(), codec.JSONReader(manifest))
 	if err != nil {
 		return nil, fmt.Errorf("failed to create request: %v", err)
 	}
+	req.Header.Set("content-type", `application/json`)
 	resp, err := s.c.Do(req)
 	if err != nil {
 		return nil, fmt.Errorf("failed to do request: %v", err)
 	}
 	defer resp.Body.Close()
 	if resp.StatusCode != http.StatusOK {
-		return nil, &clairerror.ErrRequestFail{Code: resp.StatusCode, Status: resp.Status}
+		return nil, &clairerror.ErrRequestFail{
+			Code:   resp.StatusCode,
+			Status: resp.Status,
+		}
 	}
 
-	var sr *claircore.IndexReport
-	err = json.NewDecoder(resp.Body).Decode(sr)
-	if err != nil {
-		return nil, &clairerror.ErrBadIndexReport{err}
-	}
-
-	return sr, nil
+	var ir claircore.IndexReport
+	switch ct := resp.Header.Get("content-type"); ct {
+	case "", `application/json`:
+		dec := codec.GetDecoder(resp.Body)
+		defer codec.PutDecoder(dec)
+		if err := dec.Decode(&ir); err != nil {
+			return nil, err
+		}
+	default:
+		return nil, fmt.Errorf("unrecognized content-type %q", ct)
```

Comment:
I don't know if forcing the client to provide the json content type is that useful. 

We export a JSON api, we should try to parse json, regardless of whether a client is providing a content type or not. I think we ran into this already and it was an annoyance more then a feature.

Reply:
hence ```case "", `application/json`:```

Reply:
Why the switch case then? If you're catching the default case?

Reply:
To return an error if other content-types are specified

Reply:
Right, so im saying i dont think we should restrict it at all, we have a JSON api, try to parse the JSON and return an error if invalid. Is there any harm in not being defensive here for convenience ?